lib/MySQL_Variables.cpp

```diff
@@ -0,0 +1,358 @@
+#include "MySQL_Variables.h"
+#include "proxysql.h"
+
+#include "MySQL_Session.h"
+#include "MySQL_Data_Stream.h"
+#include "SpookyV2.h"
+
+#include <sstream>
+
+MySQL_Variables::MySQL_Variables(MySQL_Session* _session) {
+	assert(_session);
+	session = _session;
+	is_connected_to_backend = false;
+
+	for (auto i = 0; i < SQL_NAME_LAST; i++) {
+		switch(i) {
+		case SQL_SAFE_UPDATES:
+		case SQL_SELECT_LIMIT:
+		case SQL_SQL_MODE:
+		case SQL_TIME_ZONE:
+		case SQL_CHARACTER_SET_RESULTS:
+		case SQL_CHARACTER_SET_CONNECTION:
+		case SQL_CHARACTER_SET_CLIENT:
+		case SQL_CHARACTER_SET_DATABASE:
+		case SQL_ISOLATION_LEVEL:
+		case SQL_TRANSACTION_READ:
+		case SQL_SESSION_TRACK_GTIDS:
+		case SQL_SQL_AUTO_IS_NULL:
+		case SQL_COLLATION_CONNECTION:
+		case SQL_NET_WRITE_TIMEOUT:
+		case SQL_MAX_JOIN_SIZE:
+			verifiers[i] = ::verify_variable;
+			updaters[i] = update_server_variable;
+			break;
+		case SQL_LOG_BIN:
+			verifiers[i] = ::verify_variable;
+			updaters[i] = logbin_update_server_variable;
+			break;
+		default:
+			updaters[i] = NULL;
+		}
+	}
+}
+
+MySQL_Variables::~MySQL_Variables() {}
+
+bool MySQL_Variables::on_connect_to_backend() {
+	if (!session || !session->mybe || !session->mybe->server_myds || !session->mybe->server_myds->myconn) return false;
+	auto be_version = session->mybe->server_myds->myconn->mysql->server_version;
+
+	if (be_version[0] == '1') {
+		int idx = SQL_NAME_LAST;
+		for (auto i=0; i<SQL_NAME_LAST; i++) {
+			if (mysql_tracked_variables[i].idx == SQL_SESSION_TRACK_GTIDS) {
+				idx = i;
+				break;
+			}
+		}
+		mysql_tracked_variables[idx].special_handling = false;
+	}
+
+	is_connected_to_backend = true;
+	return true;
+}
+
+void MySQL_Variables::client_set_value(int idx, const std::string& value) {
+	if (!session || !session->client_myds || !session->client_myds->myconn) return;
```

Comment:
A invalid usage of the function would be more difficult to debug due to being "void" but needing some internal state to hold for succeeding, maybe it would be would to add error codes to it.

Reply:
There is no internal state here. The if statement is just for safety.